pilot/pkg/networking/v1alpha3/listener.go

```diff
@@ -603,3 +620,137 @@ func buildInboundHTTPRouteConfig(instance *model.ServiceInstance) *xdsapi.RouteC
 		},
 	}
 }
+
+func buildRDSResponse(env model.Environment, node model.Proxy,
+	services []*model.Service, routeName string) *xdsapi.RouteConfiguration {
+
+	port := 0
+	if routeName != RDSHttpProxy {
+		var err error
+		port, err = strconv.Atoi(routeName)
+		if err != nil {
+			return nil
+		}
+	}
+
+	nameToServiceMap := make(map[string]*model.Service)
+	for _, svc := range services {
+		if port == 0 {
+			nameToServiceMap[svc.Hostname] = svc
+		} else {
+			if svcPort, exists := svc.Ports.GetByPort(port); exists {
+				nameToServiceMap[svc.Hostname] = &model.Service{
+					Hostname: svc.Hostname,
+					Address:  svc.Address,
+					Ports:    []*model.Port{svcPort},
+				}
+			}
+		}
+	}
+
+	guardedHosts := TranslateVirtualHosts(env.VirtualServices(), nameToServiceMap, nil, node.Domain)
```

Comment:
`VirtualServices()` implementation returns nil on error.  Does this need to smart enough to distinguish between empty array and a backend error?  Otherwise won't we send incomplete data to envoy?

Reply:
especially since we are moving to a push model for changes in services/rules instead of polling - wont this mean envoy will be stuck with the bad config until a new change is detected?